File: static/js/palette.js
// Palette Generator JavaScript

let currentPalette = [];
let lockedIndices = [];
let currentSeed = Math.floor(Math.random() * 1000000);

// Initialize
document.addEventListener('DOMContentLoaded', () => {
    setupEventListeners();
    updateSliderValues();
});

function setupEventListeners() {
    // Sliders
    const formalPlayful = document.getElementById('formal-playful');
    const modernClassic = document.getElementById('modern-classic');
    
    formalPlayful.addEventListener('input', updateSliderValues);
    modernClassic.addEventListener('input', updateSliderValues);
    
    // Generate button
    document.getElementById('generate-btn').addEventListener('click', generatePalette);
    
    // Regenerate button
    document.getElementById('regenerate-btn').addEventListener('click', regenerateUnlocked);
    
    // Expand button
    document.getElementById('expand-btn').addEventListener('click', expandPalette);
    
    // Add color input button
    document.getElementById('add-color-input').addEventListener('click', addColorInput);
    
    // Copy all hex codes button
    const copyAllBtn = document.getElementById('copy-all-hex-btn');
    if (copyAllBtn) {
        copyAllBtn.addEventListener('click', () => {
            if (currentPalette.length === 0) {
                return;
            }
            
            // Copy all hex codes as comma-separated string
            const hexCodes = currentPalette.join(', ');
            copyToClipboard(hexCodes);
        });
    }
}

function updateSliderValues() {
    const formalPlayful = document.getElementById('formal-playful').value;
    const modernClassic = document.getElementById('modern-classic').value;
    
    document.getElementById('formal-playful-value').textContent = parseFloat(formalPlayful).toFixed(2);
    document.getElementById('modern-classic-value').textContent = parseFloat(modernClassic).toFixed(2);
}

function addColorInput() {
    const container = document.getElementById('manual-colors-input');
    const input = document.createElement('input');
    input.type = 'text';
    input.className = 'color-input';
    input.placeholder = '#FFFFFF';
    input.pattern = '^#[0-9A-Fa-f]{6}$';
    container.appendChild(input);
}

function getManualColors() {
    const inputs = document.querySelectorAll('.color-input');
    const colors = [];
    inputs.forEach(input => {
        const value = input.value.trim();
        if (value && /^#[0-9A-Fa-f]{6}$/i.test(value)) {
            colors.push(value.toUpperCase());
        }
    });
    return colors;
}

function getSelectedAdjectives() {
    const checkboxes = document.querySelectorAll('.adjective-checkboxes input[type="checkbox"]:checked');
    return Array.from(checkboxes).map(cb => cb.value);
}

function generatePalette() {
    const numColors = parseInt(document.getElementById('num-colors').value);
    const formalPlayful = parseFloat(document.getElementById('formal-playful').value);
    const modernClassic = parseFloat(document.getElementById('modern-classic').value);
    const adjectives = getSelectedAdjectives();
    const manualColors = getManualColors();
    
    currentSeed = Math.floor(Math.random() * 1000000);
    
    fetch('/palette/generate', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            num_colors: numColors,
            formal_playful: formalPlayful,
            modern_classic: modernClassic,
            adjectives: adjectives,
            manual_colors: manualColors,
            seed: currentSeed
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            currentPalette = data.palette;
            lockedIndices = [];
            displayPalette(data.palette);
            document.getElementById('regenerate-btn').disabled = false;
            document.getElementById('expand-btn').disabled = false;
        } else {
            alert('Error generating palette: ' + data.error);
        }
    })
    .catch(error => {
        console.error('Error:', error);
        alert('Failed to generate palette');
    });
}

function regenerateUnlocked() {
    if (currentPalette.length === 0) return;
    
    const formalPlayful = parseFloat(document.getElementById('formal-playful').value);
    const modernClassic = parseFloat(document.getElementById('modern-classic').value);
    const adjectives = getSelectedAdjectives();
    
    currentSeed = Math.floor(Math.random() * 1000000);
    
    fetch('/palette/regenerate', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            palette: currentPalette,
            locked_indices: lockedIndices,
            formal_playful: formalPlayful,
            modern_classic: modernClassic,
            adjectives: adjectives,
            seed: currentSeed
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            currentPalette = data.palette;
            displayPalette(data.palette);
        } else {
            alert('Error regenerating palette: ' + data.error);
        }
    })
    .catch(error => {
        console.error('Error:', error);
        alert('Failed to regenerate palette');
    });
}

function expandPalette() {
    if (currentPalette.length === 0) return;
    
    const newSize = currentPalette.length + 1;
    const formalPlayful = parseFloat(document.getElementById('formal-playful').value);
    const modernClassic = parseFloat(document.getElementById('modern-classic').value);
    const adjectives = getSelectedAdjectives();
    
    currentSeed = Math.floor(Math.random() * 1000000);
    
    fetch('/palette/expand', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            palette: currentPalette,
            new_size: newSize,
            formal_playful: formalPlayful,
            modern_classic: modernClassic,
            adjectives: adjectives,
            seed: currentSeed
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            currentPalette = data.palette;
            displayPalette(data.palette);
        } else {
            alert('Error expanding palette: ' + data.error);
        }
    })
    .catch(error => {
        console.error('Error:', error);
        alert('Failed to expand palette');
    });
}

function displayPalette(palette) {
    const container = document.getElementById('palette-colors');
    const copyAllBtn = document.getElementById('copy-all-hex-btn');
    
    container.innerHTML = '';
    
    // Enable copy all button if palette exists
    if (palette.length > 0) {
        copyAllBtn.disabled = false;
    } else {
        copyAllBtn.disabled = true;
    }
    
    palette.forEach((color, index) => {
        const colorItem = document.createElement('div');
        colorItem.className = 'color-item';
        
        const isLocked = lockedIndices.includes(index);
        
        colorItem.innerHTML = `
            <div class="color-swatch" style="background-color: ${color};"></div>
            <div class="color-info">
                <span class="color-hex">${color}</span>
                <div class="color-actions">
                    <button class="lock-btn ${isLocked ? 'locked' : ''}" data-index="${index}" title="${isLocked ? 'Unlock' : 'Lock'}">
                        ${isLocked ? '🔒' : '🔓'}
                    </button>
                    <button class="copy-btn" data-color="${color}" title="Copy hex">
                        📋
                    </button>
                </div>
            </div>
        `;
        
        container.appendChild(colorItem);
    });
    
    // Add event listeners for lock and copy buttons
    container.querySelectorAll('.lock-btn').forEach(btn => {
        btn.addEventListener('click', (e) => {
            const index = parseInt(e.target.dataset.index);
            toggleLock(index);
        });
    });
    
    container.querySelectorAll('.copy-btn').forEach(btn => {
        btn.addEventListener('click', (e) => {
            const color = e.target.dataset.color;
            copyToClipboard(color);
        });
    });
}

function toggleLock(index) {
    if (lockedIndices.includes(index)) {
        lockedIndices = lockedIndices.filter(i => i !== index);
    } else {
        lockedIndices.push(index);
    }
    displayPalette(currentPalette);
}

File: blueprints/palette.py
"""
Color Palette Generator blueprint
"""

from flask import Blueprint, render_template, request, jsonify
from utils.color_utils import (
    generate_palette,
    regenerate_unlocked_colors,
    expand_palette,
    hex_to_rgb,
    rgb_to_hex
)
import json

bp = Blueprint('palette', __name__)


@bp.route('/')
def palette_generator():
    """Palette generator page"""
    return render_template('palette.html')


@bp.route('/generate', methods=['POST'])
def generate():
    """Generate a new palette based on user inputs"""
    try:
        data = request.get_json()
        
        num_colors = int(data.get('num_colors', 5))
        formal_playful = float(data.get('formal_playful', 0.5))
        modern_classic = float(data.get('modern_classic', 0.5))
        adjectives = data.get('adjectives', [])
        manual_colors = data.get('manual_colors', [])  # List of hex codes
        seed = data.get('seed', None)
        
        # If manual colors provided, use them and generate remaining
        if manual_colors:
            palette = manual_colors[:num_colors]
            remaining = num_colors - len(manual_colors)
            if remaining > 0:
                additional = generate_palette(
                    remaining, formal_playful, modern_classic, adjectives, seed
                )
                palette.extend(additional)
        else:
            palette = generate_palette(
                num_colors, formal_playful, modern_classic, adjectives, seed
            )
        
        return jsonify({
            'success': True,
            'palette': palette
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400


@bp.route('/regenerate', methods=['POST'])
def regenerate():
    """Regenerate unlocked colors in an existing palette"""
    try:
        data = request.get_json()
        
        current_palette = data.get('palette', [])
        locked_indices = data.get('locked_indices', [])
        formal_playful = float(data.get('formal_playful', 0.5))
        modern_classic = float(data.get('modern_classic', 0.5))
        adjectives = data.get('adjectives', [])
        seed = data.get('seed', None)
        
        new_palette = regenerate_unlocked_colors(
            current_palette,
            locked_indices,
            formal_playful,
            modern_classic,
            adjectives,
            seed
        )
        
        return jsonify({
            'success': True,
            'palette': new_palette
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400


@bp.route('/expand', methods=['POST'])
def expand():
    """Expand a palette to a larger size"""
    try:
        data = request.get_json()
        
        current_palette = data.get('palette', [])
        new_size = int(data.get('new_size', len(current_palette) + 1))
        formal_playful = float(data.get('formal_playful', 0.5))
        modern_classic = float(data.get('modern_classic', 0.5))
        adjectives = data.get('adjectives', [])
        seed = data.get('seed', None)
        
        expanded_palette = expand_palette(
            current_palette,
            new_size,
            formal_playful,
            modern_classic,
            adjectives,
            seed
        )
        
        return jsonify({
            'success': True,
            'palette': expanded_palette
        })
    except Exception as e:
        return jsonify({
        'success': False,
        'error': str(e)
    }), 400

